pkg/capacity/topology/nodes.go

```diff
@@ -168,6 +168,10 @@ type nodeTopology struct {
 	callbacks []Callback
 }
 
+func (nt *nodeTopology) ShutDown() {
+	nt.queue.ShutDown()
```

Comment:
Wouldn't it be better to make the code which creates the queue also shut it down? Then the Informer interface change won't be necessary.

Reply:
https://github.com/kubernetes-csi/external-provisioner/blob/0c26e9feb404a4eb807e19ded56819f0e50baa1e/pkg/capacity/topology/nodes.go#L48 the queue is passed externally to the topology informer (from the command). To shut it down from the component that creates it (in cmd), we would still need the interface change I think.

If you mean the code that starts it, This is also external as the Run code just waits for the queue to shutdown: https://github.com/kubernetes-csi/external-provisioner/blob/0c26e9feb404a4eb807e19ded56819f0e50baa1e/pkg/capacity/topology/nodes.go#L208

Maybe I misunderstood you, feel free to elaborate more if you dont think that explains why I did this

Reply:
https://github.com/kubernetes-csi/external-provisioner/blob/b377ea40edb757a7c3d8abb7058f701844ebf867/cmd/csi-provisioner/csi-provisioner.go#L474 creates the queue.

IMHO that code should be responsible for shutting it down.

Reply:
But in `cmd/csi-provisioner/csi-provisioner.go` that's pointless. If you do the same in your code, does that solve problem without any changes to this repo?

Reply:
> https://github.com/kubernetes-csi/external-provisioner/blob/b377ea40edb757a7c3d8abb7058f701844ebf867/cmd/csi-provisioner/csi-provisioner.go#L474
> 
> creates the queue.
> IMHO that code should be responsible for shutting it down.

The queue is never started with this code, so its also not responsible for shutting it down, it just owns the data structure holding the items. We start the informer in the cmd, so we also need to own the context and shut it down from what I see.

> But in cmd/csi-provisioner/csi-provisioner.go that's pointless. If you do the same in your code, does that solve problem without any changes to this repo?

In our embedded code we manually call `topologyInformer.RunWorker(ctx)` so if we expose the Shutdown method, then our code can call the Shutdown of the queue after provisioning controller shutdown and then the topologyInformer will shut down correctly as well.

Reply:
There is no explicit starting of the queue, is it?

> In our embedded code we manually call topologyInformer.RunWorker(ctx) so if we expose the Shutdown method, then our code can call the Shutdown of the queue after provisioning controller shutdown and then the topologyInformer will shut down correctly as well.

This still doesn't sound right to me. `topologyInformer.RunWorker` doesn't own any goroutines, therefore `topologyInformer.ShutDown` isn't needed. The caller can call `queue.ShutDown` instead.

Can you show your code?

Reply:
RunWorker is itself blocking so it can only reasonably called from code that uses a goroutine for it or exits on process (like you described earlier).
https://github.com/openshift/lvm-operator/blob/main/internal/csi/provisioner.go heres the code we use. Please feel free to give any suggestion how I can workaround this without having to rely on process exit and without adjusting current code.

Reply:
How about this?

```
topologyQueue := workqueue.NewRateLimitingQueueWithConfig(rateLimiter, workqueue.RateLimitingQueueConfig{
			Name: "csitopology",
		})
defer topologyQueue.ShutDown()

topologyInformer := topology.NewNodeTopology(
		p.options.DriverName,
		clientset,
		factory.Core().V1().Nodes(),
		factory.Storage().V1().CSINodes(),
		topologyQueue,
	)
```

Reply:
Okay so while this did not solve my problem directly you gave me the right path to the idea: https://github.com/openshift/lvm-operator/compare/main...jakobmoellerdev:topology-queue-shutdown?expand=1

This will work in all cases and shutdown right after the controller shuts down to avoid using the queue after shutdown and also because I have a wait group in the code.

I take from your previous comments that the other adjustments are unwanted right? Then I will close. I 100% appreciate you taking the time looking at this but also understand that we dont want to change the current system for just some context stops and logs.

Reply:
Good that you figured something out. Yes, let's keep the current code as-is.